plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php

```diff
@@ -802,61 +801,93 @@ private function months_filter() {
+	protected function get_months_filter_options(): array {
 		global $wpdb;
 
-		// We cache in the options table since it's won't be invalidated soon.
-		$cache_option_value_name = 'wc_' . $this->order_type . '_list_table_months_filter_cache_value';
-		$cache_option_date_name  = 'wc_' . $this->order_type . '_list_table_months_filter_cache_date';
-
-		$cached_timestamp = get_option( $cache_option_date_name, 0 );
-
-		// new day, new cache.
-		if ( 0 === $cached_timestamp || gmdate( 'j', time() ) !== gmdate( 'j', $cached_timestamp ) || ( time() - $cached_timestamp ) > 60 * 60 * 24 ) {
-			$cached_value = false;
-		} else {
-			$cached_value = get_option( $cache_option_value_name );
-		}
-
-		if ( false !== $cached_value ) {
-			// Always add current year month for cache stability. This allows us to not hydrate the cache on every order update.
-			$current_year_month        = new \stdClass();
-			$current_year_month->year  = gmdate( 'Y', time() );
-			$current_year_month->month = gmdate( 'n', time() );
-			if ( count( $cached_value ) === 0 || (
-				$cached_value[0]->year !== $current_year_month->year ||
-				$cached_value[0]->month !== $current_year_month->month )
-			) {
-				array_unshift( $cached_value, $current_year_month );
-			}
-			return $cached_value;
-		}
-
 		$orders_table = esc_sql( OrdersTableDataStore::get_orders_table_name() );
-		$utc_offset   = wc_timezone_offset();
+		$trash_status = esc_sql( OrderStatus::TRASH );
 
-		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
-		$order_dates = $wpdb->get_results(
+		$first_year_month = $wpdb->get_row(
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
 			$wpdb->prepare(
 				"
-					SELECT DISTINCT YEAR( t.date_created_local ) AS year,
-									MONTH( t.date_created_local ) AS month
-					FROM ( SELECT DATE_ADD( date_created_gmt, INTERVAL $utc_offset SECOND ) AS date_created_local FROM $orders_table WHERE type = %s AND status != 'trash' ) t
-					ORDER BY year DESC, month DESC
+					SELECT YEAR( t.date_created_gmt ) AS year,
+					       MONTH( t.date_created_gmt ) AS month
+					FROM $orders_table t
+					WHERE type = %s
+					AND status != '$trash_status'
+					ORDER BY year ASC, month ASC
+					LIMIT 1
 				",
 				$this->order_type
 			)
+			// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
 		);
 
-		update_option( $cache_option_date_name, time() );
-		update_option( $cache_option_value_name, $order_dates );
+		$end     = new \DateTime( 'now', wp_timezone() );
+		$options = array();
+
+		if ( is_object( $first_year_month ) ) {
+			$start = new \DateTime(
+				sprintf(
+					'%s-%s-01',
+					$first_year_month->year,
+					$first_year_month->month
+				),
+				wp_timezone()
+			);
+
+			// If, somehow, the oldest order date is in the future, swap the start and end of the range.
+			if ( $start > $end ) {
+				$end   = $start;
+				$start = new \DateTime( 'now', wp_timezone() );
+			}
+
+			while (
```

Comment:
I might be missing something, but couldn't we use `DatePeriod` instead of all this "manual" adding of intervals?

```php
foreach ( new \DatePeriod( $start, new \DateInterval( 'P1M' ), $end ) as $d ) {
	// ...
}
```

Reply:
I actually didn't know about `DatePeriod` 😳 I tried your suggestion, but it doesn't always work quite right because it excludes the end date from the list of recurrences. There's an optional flag `DatePeriod::INCLUDE_END_DATE` to adjust this, but it's not available until PHP 8.2.

Reply:
Yeah, we would have to include the end date manually, but maybe that's still an improvement / or at least more clear? This is not a blocker, of course.

Reply:
Just my personal preference, I guess, but I think it's actually clearer/less complex if we don't have to add extra logic in order to get the correct last item. I think I'll leave it for now.